cpu/cc26x0/periph/i2c.c

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright (C) 2018 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_cc26x0
+ * @ingroup     drivers_periph_i2c
+ * @{
+ *
+ * @file
+ * @brief       Low-level I2C driver implementation
+ * @note        This CPU has weak pullups, external pullup resistors may be
+ *              required.
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * @}
+ */
+
+#include <errno.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "mutex.h"
+
+#include "cpu.h"
+#include "periph/i2c.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+#define PREG(x) DEBUG("%s=0x%08x\n", #x, (unsigned)x);
+
+/**
+ * @brief Mutex lock for the only available I2C periph
+ * @note  If multiple I2C devices are added locks must be an array for each one.
+ */
+static mutex_t _lock;
+
+static int _check_errors(void)
+{
+    int ret = 0;
+
+    /* The reference manual (SWCU117H) is ambiguous on how to wait:
+     *
+     * 1. 21.4 8. says "wait until BUSBUSY is cleared"
+     * 2. command flow diagrams (e.g., 21.3.5.1) indicate to wait while
+     *    BUSY is set
+     *
+     * (3. 21.5.1.10 says BUSY is only valid after 4 SYSBUS clock cycles)
+     *
+     * Waiting first for cleared IDLE and then for cleared BUSY works fine.
+     */
+
+    /* wait for transfer to be complete, this also could be a few nops... */
+    while (I2C->MSTAT & MSTAT_IDLE) {}
+    while (I2C->MSTAT & MSTAT_BUSY) {}
+    /* check if there was an error */
+    if (I2C->MSTAT & MSTAT_ERR) {
+        DEBUG("%s\n", __FUNCTION__);
+        PREG(I2C->MSTAT);
+        ret = -ETIMEDOUT;
+        if (I2C->MSTAT & MSTAT_ADRACK_N) {
+            DEBUG("ADDRESS NACK\n");
+            return -ENXIO;
+        }
+        else if (I2C->MSTAT & MSTAT_DATACK_N) {
+            DEBUG("DATA NACK\n");
+            ret = -EIO;
+        }
+        else if (I2C->MSTAT & MSTAT_ARBLST) {
+            DEBUG("ARBITRATION LOSS\n");
+            ret = -EAGAIN;
+        }
+        /*
+         * If an error if a non-NACK error occurs we must reinit or get stuck.
```

Comment:
sentence starts weird

Reply:
Heh yup, race condition in my head I guess.